utils/pkcs11helpers.go

```diff
@@ -0,0 +1,430 @@
+		MGF:        pkcs11.CKG_MGF1_SHA1,
+		SourceType: pkcs11.CKZ_DATA_SPECIFIED,
+		SourceData: nil,
+	}
+	// OAEPSha1Params describes the OAEP parameters with sha256 hash algorithm
+	OAEPSha256Params = &pkcs11.OAEPParams{
+		HashAlg:    pkcs11.CKM_SHA256,
+		MGF:        pkcs11.CKG_MGF1_SHA256,
+		SourceType: pkcs11.CKZ_DATA_SPECIFIED,
+		SourceData: nil,
+	}
+	// OEPParams is an array Of OAEP parameters in order we will try to use them
+	OAEPParams = []*pkcs11.OAEPParams{OAEPSha256Params, OAEPSha1Params}
+)
+
+// Pkcs11KeyFile describes the format of the pkcs11 (private) key file
+type Pkcs11KeyFile struct {
+	Pkcs11 struct {
+		Uri string `yaml:"uri"`
+	} `yaml:"pkcs11"`
+}
+
+// Pkcs11KeyFileObject is a representation of the Pkcs11KeyFile with the pkcs11 URI as an object
+type Pkcs11KeyFileObject struct {
+	uri pkcs11uri.Pkcs11URI
+}
+
+// ParsePkcs11KeyFile parses a pkcs11 key file holding a pkcs11 URI describing a private key.
+// The file has the following yaml format:
+// pkcs11:
+//  - uri : <pkcs11 uri>
+// An error is returned if the pkcs11 URI is malformed
+func ParsePkcs11KeyFile(yamlstr []byte) (*Pkcs11KeyFileObject, error) {
+	p11keyfile := Pkcs11KeyFile{}
+
+	err := yaml.Unmarshal([]byte(yamlstr), &p11keyfile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not unmarshal pkcs11 keyfile")
+	}
+
+	p11uri, err := pkcs11uri.New()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not create Pkcs11URI object")
+	}
+	err = p11uri.Parse(p11keyfile.Pkcs11.Uri)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not parse Pkcs11URI from file")
+	}
+
+	return &Pkcs11KeyFileObject{uri: p11uri}, err
+}
+
+// Pkcs11Config describes the layout of a pkcs11 config file
+// The file has the following yaml format:
+// module-directories:
+// - /usr/lib64/pkcs11/
+// allowd-module-paths
+// - /usr/lib64/pkcs11/libsofthsm2.so
+type Pkcs11Config struct {
+	ModuleDirectories  []string `yaml:"module-directories"`
+	AllowedModulePaths []string `yaml:"allowed-module-paths"`
+}
+
+// ParsePkcs11ConfigFile parses a pkcs11 config file hat influences the module search behavior
+// as well as the set of modules that users are allowed to use
+func ParsePkcs11ConfigFile(yamlstr []byte) (Pkcs11Config, error) {
+	p11conf := Pkcs11Config{}
+
+	err := yaml.Unmarshal([]byte(yamlstr), &p11conf)
+	if err != nil {
+		return p11conf, errors.Wrapf(err, "Could not parse Pkcs11Config")
+	}
+	return p11conf, nil
+}
+
+// rsaPublicEncryptOAEP encrypts the given plaintext with the given *rsa.PublicKey; the
+// environment variable OCI_OAEP_HASHALG can be set to 'sha1' to force usage of sha1 for OAEP (SoftHSM).
+// This function is needed by clients who are using a public key file for pkcs11 encryption
+func rsaPublicEncryptOAEP(pubKey *rsa.PublicKey, plaintext []byte) ([]byte, error) {
+	var hashfunc hash.Hash
+
+	// for SoftHSM support we allow the user to choose sha1 as the hash algorithm
+	switch strings.ToLower(os.Getenv("OCI_OAEP_HASHALG")) {
```

Comment:
@lumjjb  `ECI_OAEP_HASHALG` or `OCICRYPT_OAEP_HASHALG` -- maybe either one of those would be better?

Reply:
Hmm, yea maybe `OCICRYPT_OAEP_HASHALG`

Reply:
Done that now.